parsers/ZA.py

```diff
@@ -0,0 +1,965 @@
+        2: "unknown",  # Eskom_Gas_SCO_Pumping              changed to unknown since negative gas is not possible
+        3: "hydro",  # Hydro_Water_SCO_Pumping
+        4: "hydro",  # Pumped_Water_SCO_Pumping
+        5: "unknown",  # Thermal_Generation                 sum of 0, 1, 2, 3, 4. This value is not used.
+        6: "nuclear",  # Nuclear_Generation
+        7: "unknown",  # International_Imports
+        8: "oil",  # Eskom_OCGT_Generation
+        9: "gas",  # Eskom_Gas_Generation
+        10: "oil",  # Dispatchable_IPP_OCGT
+        11: "hydro",  # Hydro_Water_Generation
+        12: "hydro",  # Pumped_Water_Generation
+        13: "unknown",  # IOS_Excl_ILS_and_MLR              Interruption of Supply. Not used.
+        14: "unknown",  # ILS_Usage                         Interruptible Load Shed = companies paid not to consume electricity. Not used.
+        15: "unknown",  # Manual_Load_Reduction_MLR         MLS = forced load shedding. Not used.
+        16: "wind",  # Wind
+        17: "solar",  # PV
+        18: "solar",  # CSP
+        19: "unknown",  # Other_RE
+    }
+
+    all_data = []
+
+    for i, (datetime_key, value) in enumerate(merged_csv_bi_data.items()):
+
+        # The production values (MW) should never be negative. Use None, or omit the key if # a specific production mode is not known.
+        # ---
+        # storage values can be both positive (when storing energy) or negative (when the
+        # storage is discharged).
+
+        data = {
+            "zoneKey": zone_key,
+            "datetime": datetime_key.datetime,
+            "production": {
+                "biomass": None,
+                "coal": 0.0,
+                "gas": 0.0,
+                "hydro": 0.0,
+                "nuclear": 0.0,
+                "oil": 0.0,
+                "solar": 0.0,
+                "wind": 0.0,
+                "geothermal": None,
+                "unknown": 0.0,
+            },
+            "storage": {"hydro": 0.0},
+            "source": " https://www.eskom.co.za",
+        }
+
+        cleansed_csv_values = value["cleansed_csv"].split(",")
+
+        storage_inversion_idcs = [3, 4, 12]
+        production_idcs = [0, 1, 2, 6, 8, 9, 10, 11, 16, 17, 18, 19]
+
+        # Omitted values
+        # 5, 7, 13, 14, 15
+        # TODO:
+        # - 7 (international imports) can be further implemented in exchange function.
+
+        for j, cleansed_csv_value in enumerate(cleansed_csv_values):
+            if j in storage_inversion_idcs:
+                data["storage"][column_mapping[j]] = round(
+                    data["storage"][column_mapping[j]] + float(cleansed_csv_value) * -1,
+                    2,
+                )
+            elif j in production_idcs:
+                data["production"][column_mapping[j]] = round(
+                    data["production"][column_mapping[j]] + float(cleansed_csv_value),
+                    2,
+                )
+
```

Comment:
```suggestion
        for j, cleansed_csv_value in enumerate(cleansed_csv_values):
            if j in storage_inversion_idcs:
                data["storage"][column_mapping[j]] = round(
                    data["storage"][column_mapping[j]] + float(cleansed_csv_value) * -1,
                    13,
                )
            elif j in production_idcs:
                data["production"][column_mapping[j]] = round(
                    data["production"][column_mapping[j]] + float(cleansed_csv_value),
                    13,
                )

```
Reverting this change as it reduces the data resolution, I didn't think 13 decimals was needed but the general consensus seem to be that we should keep the data as close to the source as possible.

Reply:
Done in commit 024737e2f1e4c96773dd810f06a60fd479b69ecb